### PR TITLE
refactor: swap `bernoulli` and `bernoulli'`

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
i.e. `bernoulli 1 = +1 / 2` and `bernoulli' 1 = -1 / 2`. Defining `bernoulli 1 = +1 / 2` makes more formulas involving Bernoulli numbers simpler; see `riemannZeta_neg_nat_eq_bernoulli` for an example and
* https://github.com/sympy/sympy/issues/23866
* https://github.com/sympy/sympy/pull/23984

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/Bernoulli.lean [hunk 1]
```diff
@@ -187,97 +183,96 @@ theorem bernoulli'_odd_eq_zero {n : ℕ} (h_odd : Odd n) (hlt : 1 < n) : bernoul
...
-  simp [bernoulli, ← mul_assoc, ← sq, ← pow_mul, mul_comm n 2, pow_mul]
-#align bernoulli'_eq_bernoulli bernoulli'_eq_bernoulli
+theorem bernoulli_eq_bernoulli' (n : ℕ) : bernoulli n = (-1) ^ n * bernoulli' n := by
+  simp [bernoulli', ← mul_assoc, ← sq, ← pow_mul, mul_comm n 2, pow_mul]
+#align bernoulli'_eq_bernoulli bernoulli_eq_bernoulli'
 
 @[simp]
-theorem bernoulli_zero : bernoulli 0 = 1 := by simp [bernoulli]
+theorem bernoulli'_zero : bernoulli' 0 = 1 := by simp [bernoulli']
 #align bernoulli_zero bernoulli_zero
```
Comment (Parcly-Taxel) [inline Mathlib/NumberTheory/Bernoulli.lean:205]:
```suggestion
#align bernoulli_zero bernoulli'_zero
```

**Review by loefflerd — COMMENTED** [review]
I think this is a bad idea. The Wikipedia article explicitly writes that the $B_1 = -1/2$ convention "is the sign convention prescribed by NIST and most modern textbooks", and checking the books on my shelf, e.g. Washington's *Cyclotomic Fields*, confirms this.

**Comment by Parcly-Taxel** [comment]
> "is the sign convention prescribed by NIST and most modern textbooks"

I know. But as Peter Luschny (hence this branch's name) writes in his "Bernoulli Manifesto" $B_1=-1/2$ is wrong on multiple levels, some of which follow:
* Bernoulli's _Ars Conjectandi_ implies $B_1=+1/2$, though of course he never thought of his numbers as a sequence.
* The Euler–Maclaurin formula is much simpler with $B_1=+1/2$.
* `riemannZeta_neg_nat_eq_bernoulli` needs no awkward $(-1)^n$ term if $B_1=+1/2$.

Luschny convinced Knuth to correct to $B_1=+1/2$ in the 2021 version of _Concrete Mathematics_.

**Comment by loefflerd** [comment]
I’m sorry, but I don’t consider any of those arguments sufficient to outweigh the obvious disadvantage of making Mathlib incompatible with the bulk of the existing unformalised literature on the topic.

**Comment by jcommelin** [comment]
How do people think about `bernouilli₋` and `bernouilli₊` as names for the two variants? Then we don't have to claim a preference to begin with.

**Comment by Parcly-Taxel** [comment]
> I’m sorry, but I don’t consider any of those arguments sufficient to outweigh the obvious disadvantage of making Mathlib incompatible with the bulk of the existing unformalised literature on the topic.

Incompatibility with existing literature is a non-issuee. We, as a formal mathematics library built from (essentially) scratch, have no obligations to literature; we can choose definitions to maximise ease of theorem stating and proving, and $B_1=+1/2$ goes more in this direction than $B_1=-1/2$ as Luschny found out.

**Comment by jcommelin** [comment]
> > I’m sorry, but I don’t consider any of those arguments sufficient to outweigh the obvious disadvantage of making Mathlib incompatible with the bulk of the existing unformalised literature on the topic.
> 
> Incompatibility with existing literature is a non-issue. We, as a formal mathematics library built from (essentially) scratch, have no obligations to literature

Meh, we don't have any obligations, but we sure care about compatibility to some extent. We could rename `Group` to `g`, and many statements would become shorter. But we don't do that.

This design choice is well-known within maths, and conventions are in place.

Following the mainstream convention is a very reasonable argument.

**Comment by Parcly-Taxel** [comment]
> Following the mainstream convention is a very reasonable argument.

I don't think a mainstream convention exists anymore after the change in GKP (_Concrete Mathematics_). IIRC @teorth and @fredrik-johansson (maintainer of the mpmath library that SymPy depends on) also hold that $B_1=+1/2$.

Luschny's argument essentially boils down to this: the A&S/NIST DLMF/Mathematica preference of $B_1=-1/2$ obscures the deep relationship between Bernoulli numbers, sums of powers and zeta functions in a way that $B_1=+1/2$ does not. Who wouldn't want a deeper understanding of things?

**Comment by loefflerd** [comment]
@Parcly-Taxel Your comments on the formalisation of Riemann zeta in mathlib are sadly misinformed.

The argument that I formalised to prove the analytic continuation – which is in fact taken essentially verbatim from Riemann's original paper _On the Number of Primes Less Than a Given Magnitude_ – is not the simplest way of proving the analytic continuation, alone. However, it has the advantage that it also gives the functional equation relating s and 1 - s, which is a much deeper result, and (unless I'm much mistaken) would be very much harder to derive from elementary methods such as the sum formula of Hasse you quote. (Euler knew, in some form, the analytic continuation; the functional equation was a new result in Riemann's paper.) 

The argument, as formalised, also extends to other modular forms: the Riemann zeta function is (up to a minor substitution) the L-series of the Jacobi theta series, which is a modular form of half-integer weight, and one can drop in any other modular form, and obtain analytic continuation and functional equations for all modular form L-functions with very little modification.

**Comment by kbuzzard** [comment]
I was asked by a maintainer if I had an opinion on this. The original work on Bernoulli numbers was done by Ashvni Narayanan, a PhD student of mine, and we did wrestle with this question, trying to decide which of the choices was "the right one". Let me tell you the observation which made us give up.

The formula for the sum of the first n k'th powers is a polynomial of degree k+1 in n, and the beautiful formula for it (although it doesn't look so beautiful in plain text) is 

(1/(k+1))*((k+1 choose 0) B_0 n^{k+1} + (k+1 choose 1) B_1 n^k + (k+1 choose 2) B_2 n^{k-1}+ ...)

= n^{k+1}/(k+1) + B_1 n^k + (k/2) B_2 n^{k-1} + ...

where B_n is the n'th Bernoulli number. This struck me as such a beautiful formula that we might want to be informed by it. But then I realised that if by "the first n k'th powers" you mean 1^k+2^k+...+n^k, then you get B_1=+1/2, and if you mean 0^k+1^k+...+(n-1)^k then you obviously lose a term n^k so you get B_1=-1/2. Put it another way, are we summing i^k over i<n or i<=n? There is no "right" or "wrong" or "canonical" answer to this question; < and <= are two equivalent ways of packaging up the concept of a partial order; 0 and 1 are both perfectly good places to start when considering questions about counting, and there seemed to me to be absolutely no way of saying which one was "right". Figuring out which one is "right" is like trying to figure out whether it's "right" to start the natural numbers at 0 or 1, this is just basically a meaningless question; sometimes it's better to start at 0, and sometimes it's better to start at 1, it just depends on what you're doing.

So in conclusion I think we should either do nothing (because following the literature is as good a reason as any, given that there is no "right" answer tp override this), or follow Johan's suggestion (i.e. follow Wikipedia) and use + and - as distinguishing markers. I don't buy the fact that Knuth changed his mind as being evidence that one of the choices is "right" -- indeed I am arguing that there is no possible choice which is right for everyone, and what is right for Knuth (who is interested in counting) might not be right for an analyst (who is interested in the zeta function) or for someone else. 

In my own field there is a similar question of whether the arithmetic Frobenius or the geometric Frobenius is "right" (they're elements of Galois groups and one is the inverse of the other), and again I've come to the same conclusion: people who work in one sub-area of the field have a preference for a particular choice because it minimises the amount of minus signs which occur in the formulas which they personally are interested in; but there is no global "right" answer.

**Comment by loefflerd** [comment]
I'm not keen on the `bernoulli₋` and` bernoulli₊` option: it's inviting a scenario where @Parcly-Taxel opens up a separate PR for every single Bernoulli-related lemma in the library, expounding at length why his preferred option is obviously the best one. 

This tedious debate has already consumed more energy and attention than it deserves. (If this Luschny chap has indeed spent a decade pushing for a particular sign convention, then I pity him for not having anything better to do.) Let's settle it now, one way or the other.

I have a slight preference for the "-" over the "+" convention, but I have a much, much stronger preference for having *some* convention over complete anarchy

**Pull request closed without merging**